design-docs/Defer merging implementation.md

```diff
@@ -0,0 +1,445 @@
+  }
+}
+```
+
+And with the path `["computers", 0]`, would result in:
+
+```json
+{
+  "computers": [
+    {
+      "__typename": "Computer",
+      "id": "Computer1",
+      "cpu": "386",
+      "year": 1993,
+      "screen": {
+        "resolution": "640x480"
+      }
+    },
+    {
+      "__typename": "Computer",
+      "id": "Computer2"
+    }
+  ]
+}
+```
+
+Converting the JSONs to `Map<String, Any?>` should make the implementation of such function not too difficult.
+
+#### Adapters must parse deferred fragments only if present
+
+When parsing payload 1, the fields of the `ComputerFields` fragment are not present, so the adapter should not try to
+read them.
+
+However, when parsing "payload 1 merged with payload 2", the fields for `ComputerFields` are present and must be read.
+
+We need a way to switch this reading on/off depending on whether specific fragments have been merged or not.
+
+Solution:
+
+- Like with Approach A, add a unique label to each `@defer` directive.
+- Add a property `deferredFragmentLabels: Set<String>` to `CustomScalarAdapters`
+    - ⚠️ This is kind of a hack - we hijack the original purpose of `CustomScalarAdapters` and use it as a kind of
+      context. This avoids changing the Adapter API in a breaking way.
+    - We already used this technique to pass around execution variables, which are needed at parse time for `@skip`
+      / `@include`.
+- When handling a payload, add its label to `customScalarAdapters.deferredFragmentLabels`.
+- The Adapter's generated code only parses a fragment marked with `@defer` if the associated label is present
+  in `customScalarAdapters.deferredFragmentLabels`.
+
+For instance currently the generated code for `ComputerFieldsImpl_ResponseAdapter.ComputerFields` looks like this:
+
+```kotlin
+reader.rewind()
+val screenFields = com.example.fragment.ScreenFieldsImpl_ResponseAdapter.ScreenFields.fromJson(reader,
+    customScalarAdapters)
+
+return com.example.fragment.ComputerFields.Screen(
+    __typename = __typename,
+    resolution = resolution!!,
+    screenFields = screenFields
+)
+```
+
+Instead, we'd have:
+
+```kotlin
+reader.rewind()
+var screenFields: ScreenFields? = null
+if (customScalarAdapters.deferredFragmentLabels.contains("fragment:ComputerFields:0")) {
+    screenFields = com.example.fragment.ScreenFieldsImpl_ResponseAdapter.ScreenFields.fromJson(reader,
+        customScalarAdapters)
+}
+
+return com.example.fragment.ComputerFields.Screen(
+    __typename = __typename,
+    resolution = resolution!!,
+    screenFields = screenFields
+)
+```
+
+### Comparison of both approaches
+
+Approach A:
+
+**Pros:**
+- No need to parse the JSON into Maps
+- Each payload is parsed only once
+
+**Cons:**
+- Fairly complex / difficult (if even possible?) code generation to merge data classes
+
+
+Approach B:
+
+**Pros:**
+- Changes to the generated Adapter are minimal
+
+**Cons:**
+- JSON payloads must be parsed into Maps, which are kept in memory for a while
```

Comment:
Technically, that's better than keep the data classes, which may contain duplicated merged fields, in memory.